File: back/src/main/java/csh/back/domain/trip/group/entity/TripGroup.java
package csh.back.domain.trip.group.entity;

import csh.back.domain.member.entity.Member;
import csh.back.domain.trip.group.dto.request.TripGroupModifyRequest;
import csh.back.domain.trip.member.entity.TripMember;
import csh.back.global.entity.BaseEntity;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.ColumnDefault;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;


//tripGroups 엔티티
@Getter
@Entity
@Table(name = "tripGroups")
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class TripGroup extends BaseEntity {
	//방장
	@ManyToOne(fetch = FetchType.LAZY)
	//FK
	//join member Table
	@JoinColumn(name = "member_id", nullable = false)
	private Member owner;

	//여행 그룹 이름
	private String name;

	//지역 이름
	private String region;

	//nights
	private int nights;

	//초대 링크
	@Column(unique = true)
	private String joinCode;

	//여행 시작일
	private LocalDate startDate;

	//여행 마무리 날짜
	private LocalDate endDate;

	//투표 시작 상태(초기 생성은 false)
	@ColumnDefault("false")
	private boolean isVote;

	//생성자
	//buider 사용
	@Builder
	private TripGroup(Member owner, String name, String region, int nights, String joinCode, LocalDate startDate, LocalDate endDate) {
		this.owner = owner;
		this.name = name;
		this.region = region;
		this.nights = nights;
		this.joinCode = joinCode;
		this.startDate = startDate;
		this.endDate = endDate;
	}

	//name 수정 메서드
	public void modify(TripGroupModifyRequest request) {
		if (request.name() != null && !request.name().isBlank()) this.name = request.name();
		if (request.region() != null && !request.region().isBlank()) this.region = request.region();
		if (request.startDate() != null && !request.startDate().isBlank()) {
			// endDate를 받는게 아니라면 startDate가 변할때 endDate도 수정해줘야 한다.
			this.startDate = LocalDate.parse(request.startDate());
			this.endDate = this.startDate.plusDays(this.nights);
		}
		if (request.nights() != null) {
			this.nights = request.nights();
			this.endDate = this.startDate.plusDays(request.nights());
		}
	}
}

File: back/src/main/kotlin/csh/back/domain/trip/group/entity/TripGroup.kt
package csh.back.domain.trip.group.entity

import csh.back.domain.member.entity.Member
import csh.back.domain.trip.group.dto.request.TripGroupModifyRequest
import csh.back.global.entity.BaseEntity
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.FetchType
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.Table
import org.hibernate.annotations.ColumnDefault
import org.hibernate.annotations.SQLDelete
import org.hibernate.annotations.SQLRestriction
import java.time.LocalDate
import java.time.LocalDateTime

// Soft delete:
//   @SQLDelete: repository.delete(group) 호출 시 실제 DELETE 대신 UPDATE로 deleted_at을 채움.
//   @SQLRestriction: JPA 조회에 자동으로 WHERE deleted_at IS NULL 을 붙임.
//     주의: QueryDSL 쿼리에는 자동 적용되지 않으므로 TripGroupRepositoryImpl에서 명시적으로 조건을 넣어야 함.
@Entity
@Table(name = "trip_groups")
@SQLDelete(sql = "UPDATE trip_groups SET deleted_at = CURRENT_TIMESTAMP WHERE id = ?")
@SQLRestriction("deleted_at IS NULL")
class TripGroup(
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "member_id", nullable = false)
    val owner: Member,

    var name: String,
    var region: String,
    var nights: Int,

    @Column(unique = true)
    val joinCode: String,

    var startDate: LocalDate,
    var endDate: LocalDate,
) : BaseEntity() {

    @ColumnDefault("false")
    var isVote: Boolean = false

    // @SQLDelete가 UPDATE로 채워주는 컬럼. null이면 살아있는 방, 값이 있으면 삭제된 방.
    // 애플리케이션 코드에서 이 필드를 직접 세팅하지 말고 repository.delete()를 통해서만 세팅해야
    // @SQLDelete의 자동 처리와 일관성이 유지됨.
    var deletedAt: LocalDateTime? = null

    fun modify(request: TripGroupModifyRequest) {
        if (!request.name.isNullOrBlank()) this.name = request.name
        if (!request.region.isNullOrBlank()) this.region = request.region
        if (!request.startDate.isNullOrBlank()) {
            this.startDate = LocalDate.parse(request.startDate)
            this.endDate = this.startDate.plusDays(this.nights.toLong())
        }
        if (request.nights != null) {
            this.nights = request.nights
            this.endDate = this.startDate.plusDays(request.nights.toLong())
        }
    }
}
